### ISSUE TITLE
mAP in RetinaNet goes to 0 after first evaluation and all losses went up

### ISSUE THREAD

**Issue by sarmientoj24** [open-mmlab/mmdetection#4595]
Upon using `retinanet_r50_fpn_fp16_1x_coco`, after the first evaluation that went well with mAP as high as `0.392`, mAP goes to 0 after every evaluation and losses went up. _**Learning rate seems to also fluctuate in the beginning**_

**First mAP evaluation**
```
---------------iou_thr: 0.5---------------
2021-02-05 17:17:44,230 - mmdet - INFO - 
+-------+------+--------+--------+-------+
| class | gts  | dets   | recall | ap    |
+-------+------+--------+--------+-------+
| D00   | 1348 | 110751 | 0.921  | 0.355 |
| D10   | 958  | 67876  | 0.864  | 0.243 |
| D20   | 1739 | 84627  | 0.946  | 0.582 |
| D40   | 1137 | 76872  | 0.872  | 0.390 |
+-------+------+--------+--------+-------+
| mAP   |      |        |        | 0.392 |
+-------+------+--------+--------+-------+
2021-02-05 17:18:22,858 - mmdet - INFO - Exp name: mmretinanet.py
2021-02-05 17:18:22,859 - mmdet - INFO - Epoch(val) [5][78]	loss_cls: 0.3732, loss_bbox: 0.3992, loss: 0.7724
```

**Second mAP evaluation**
```
---------------iou_thr: 0.5---------------
2021-02-05 17:41:53,254 - mmdet - INFO - 
+-------+------+------+--------+-------+
| class | gts  | dets | recall | ap    |
+-------+------+------+--------+-------+
| D00   | 1348 | 0    | 0.000  | 0.000 |
| D10   | 958  | 0    | 0.000  | 0.000 |
| D20   | 1739 | 0    | 0.000  | 0.000 |
| D40   | 1137 | 0    | 0.000  | 0.000 |
+-------+------+------+--------+-------+
| mAP   |      |      |        | 0.000 |
+-------+------+------+--------+-------+
2021-02-05 17:42:31,574 - mmdet - INFO - Exp name: mmretinanet.py
2021-02-05 17:42:31,575 - mmdet - INFO - Epoch(val) [10][78]	loss_cls: 21.8341, loss_bbox: 0.6821, loss: 22.5162
```

**custom.py**
```
# model settings
fp16 = dict(loss_scale=512.)
model = dict(
    type='RetinaNet',
    pretrained='torchvision://resnet50',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=True,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        start_level=1,
        add_extra_convs='on_input',
        num_outs=5),
    bbox_head=dict(
        type='RetinaHead',
        num_classes=80,
        in_channels=256,
        stacked_convs=4,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            octave_base_scale=4,
            scales_per_octave=3,
            ratios=[0.5, 1.0, 2.0],
            strides=[8, 16, 32, 64, 128]),
        bbox_coder=dict(
            type='DeltaXYWHBBoxCoder',
            target_means=[.0, .0, .0, .0],
            target_stds=[1.0, 1.0, 1.0, 1.0]),
        loss_cls=dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),
        loss_bbox=dict(type='L1Loss', loss_weight=1.0)),
    # training and testing settings
    train_cfg=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.4,
            min_pos_iou=0,
            ignore_iof_thr=-1),
        allowed_border=-1,
        pos_weight=-1,
        debug=False),
    test_cfg=dict(
        nms_pre=1000,
        min_bbox_size=0,
        score_thr=0.05,
        nms=dict(type='nms', iou_threshold=0.5),
        max_per_img=100))

dataset_type = 'VOCDataset'
data_root = 'VOC2012/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(512, 512), keep_ratio=True),
    dict(type='Pad', size_divisor=32),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(512, 512),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='Pad', size_divisor=32),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=32,
    workers_per_gpu=2,
    train=dict(
        type='VOCDataset',
        ann_file='VOC2012/ImageSets/Main/train.txt',
        img_prefix='VOC2012/',
        pipeline=train_pipeline,
        classes=('D00', 'D10', 'D20', 'D40')),
    val=dict(
        type='VOCDataset',
        ann_file='VOC2012/ImageSets/Main/val.txt',
        img_prefix='VOC2012/',
        pipeline=test_pipeline,
        classes=('D00', 'D10', 'D20', 'D40')),
    test=dict(
        type='VOCDataset',
        ann_file='VOC2012/ImageSets/Main/test.txt',
        img_prefix='VOC2012/',
        pipeline=test_pipeline,
        classes=('D00', 'D10', 'D20', 'D40'),
        samples_per_gpu=2))

optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=None)
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=0.001,
    step=[8, 11])

total_epochs = 150
checkpoint_config = dict(interval=5)
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TensorboardLoggerHook'),
        dict(type='TextLoggerHook'),
        dict(
            type='WandbLoggerHook',
            init_kwargs=dict(project='mmdetection', name='retina-net-150'))
    ])
custom_hooks = [dict(type='NumClassCheckHook')]
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = '/mmdetection/checkpoint/retinanet_r50_fpn_fp16_1x_coco_20200702-0dbfb212.pth'
resume_from = None
workflow = [('train', 1), ('val', 1)]
classes = ('D00', 'D10', 'D20', 'D40')
work_dir = './exps'
gpu_ids = range(0, 1)
evaluation = dict(interval=5, metric='mAP')
```

**Wandb chart of losses**
![image](https://user-images.githubusercontent.com/8830319/107115808-18848880-68aa-11eb-8007-d2069a4c4121.png)

**Learning rate graph**
![image](https://user-images.githubusercontent.com/8830319/107115907-cabc5000-68aa-11eb-8dcf-3ae1bdb11cc8.png)

**First few lines of log**
```
2021-02-05 16:54:36,834 - mmdet - INFO - Epoch [1][50/304]	lr: 9.890e-04, eta: 10:06:47, time: 0.799, data_time: 0.141, memory: 5695, loss_cls: 0.9992, loss_bbox: inf, loss: inf
2021-02-05 16:55:11,534 - mmdet - INFO - Epoch [1][100/304]	lr: 1.988e-03, eta: 9:26:12, time: 0.694, data_time: 0.044, memory: 5695, loss_cls: 0.6690, loss_bbox: 0.5139, loss: 1.1828
2021-02-05 16:55:46,210 - mmdet - INFO - Epoch [1][150/304]	lr: 2.987e-03, eta: 9:12:09, time: 0.693, data_time: 0.045, memory: 5695, loss_cls: 0.5367, loss_bbox: 0.4643, loss: 1.0009
2021-02-05 16:56:20,848 - mmdet - INFO - Epoch [1][200/304]	lr: 3.986e-03, eta: 9:04:42, time: 0.693, data_time: 0.044, memory: 5695, loss_cls: 0.4852, loss_bbox: 0.4438, loss: 0.9290
2021-02-05 16:56:55,799 - mmdet - INFO - Epoch [1][250/304]	lr: 4.985e-03, eta: 9:00:57, time: 0.699, data_time: 0.049, memory: 5695, loss_cls: 0.4603, loss_bbox: 0.4436, loss: 0.9038
2021-02-05 16:57:30,554 - mmdet - INFO - Epoch [1][300/304]	lr: 5.984e-03, eta: 8:57:45, time: 0.695, data_time: 0.045, memory: 5695, loss_cls: 0.4431, loss_bbox: 0.4285, loss: 0.8716
2021-02-05 16:58:12,116 - mmdet - INFO - Exp name: mmretinanet.py
2021-02-05 16:58:12,116 - mmdet - INFO - Epoch(val) [1][78]	loss_cls: 0.4479, loss_bbox: 0.4258, loss: 0.8737
2021-02-05 16:58:51,877 - mmdet - INFO - Epoch [2][50/304]	lr: 7.063e-03, eta: 8:59:49, time: 0.795, data_time: 0.137, memory: 5695, loss_cls: 0.4242, loss_bbox: 0.4211, loss: 0.8453
2021-02-05 16:59:26,666 - mmdet - INFO - Epoch [2][100/304]	lr: 8.062e-03, eta: 8:57:21, time: 0.696, data_time: 0.044, memory: 5695, loss_cls: 0.4412, loss_bbox: 0.4245, loss: 0.8657
2021-02-05 17:00:01,484 - mmdet - INFO - Epoch [2][150/304]	lr: 9.061e-03, eta: 8:55:21, time: 0.696, data_time: 0.045, memory: 5695, loss_cls: 0.4294, loss_bbox: 0.4194, loss: 0.8488
2021-02-05 17:00:36,484 - mmdet - INFO - Epoch [2][200/304]	lr: 1.000e-02, eta: 8:53:54, time: 0.700, data_time: 0.045, memory: 5695, loss_cls: 0.3994, loss_bbox: 0.4136, loss: 0.8130
2021-02-05 17:01:11,461 - mmdet - INFO - Epoch [2][250/304]	lr: 1.000e-02, eta: 8:52:34, time: 0.700, data_time: 0.048, memory: 5695, loss_cls: 0.4315, loss_bbox: inf, loss: inf
2021-02-05 17:01:46,290 - mmdet - INFO - Epoch [2][300/304]	lr: 1.000e-02, eta: 8:51:11, time: 0.697, data_time: 0.045, memory: 5695, loss_cls: 0.4029, loss_bbox: 0.4162, loss: 0.8191
2021-02-05 17:02:27,545 - mmdet - INFO - Exp name: mmretinanet.py
2021-02-05 17:02:27,545 - mmdet - INFO - Epoch(val) [2][78]	loss_cls: 0.3961, loss_bbox: 0.4102, loss: 0.8062
2021-02-05 17:03:06,900 - mmdet - INFO - Epoch [3][50/304]	lr: 1.000e-02, eta: 8:51:46, time: 0.787, data_time: 0.124, memory: 5695, loss_cls: 0.3740, loss_bbox: 0.3917, loss: 0.7657
2021-02-05 17:03:41,739 - mmdet - INFO - Epoch [3][100/304]	lr: 1.000e-02, eta: 8:50:29, time: 0.697, data_time: 0.046, memory: 5695, loss_cls: 0.3718, loss_bbox: 0.3927, loss: 0.7646
2021-02-05 17:04:16,619 - mmdet - INFO - Epoch [3][150/304]	lr: 1.000e-02, eta: 8:49:20, time: 0.698, data_time: 0.045, memory: 5695, loss_cls: 0.3682, loss_bbox: 0.3961, loss: 0.7643
2021-02-05 17:04:51,479 - mmdet - INFO - Epoch [3][200/304]	lr: 1.000e-02, eta: 8:48:14, time: 0.697, data_time: 0.045, memory: 5695, loss_cls: 0.3725, loss_bbox: 0.3927, loss: 0.7652
2021-02-05 17:05:26,494 - mmdet - INFO - Epoch [3][250/304]	lr: 1.000e-02, eta: 8:47:19, time: 0.700, data_time: 0.047, memory: 5695, loss_cls: 0.3979, loss_bbox: inf, loss: inf
2021-02-05 17:06:01,436 - mmdet - INFO - Epoch [3][300/304]	lr: 1.000e-02, eta: 8:46:23, time: 0.699, data_time: 0.046, memory: 5695, loss_cls: 0.3724, loss_bbox: 0.3931, loss: 0.7655
2021-02-05 17:06:43,024 - mmdet - INFO - Exp name: mmretinanet.py
2021-02-05 17:06:43,024 - mmdet - INFO - Epoch(val) [3][78]	loss_cls: 0.3861, loss_bbox: 0.4072, loss: 0.7934
```

Is there any particularly wrong here?

**Comment by jshilong** [comment]
Obviously,  there are abnormal values in your log, such as inf, Focal loss is sensitive for the noisy data，I recommend you add `gradclip` in your config to stable the training process.

**Comment by sarmientoj24** [comment]
@jshilong I just used what is available for RetinaNet. Is there an example for this? What would be okay for default gradient clipping?

**Comment by jshilong** [comment]
Sure, you can refer to the same setting in `fsaf`:
https://github.com/open-mmlab/mmdetection/blob/272f2007397d8aa984195d2cc6ac6710b8d1fd5d/configs/fsaf/fsaf_r50_fpn_1x_coco.py#L47
